src/WatchDogs.Domain/SuspiciousDealDetector.cs

```diff
@@ -0,0 +1,279 @@
+    public ConcurrentDictionary<string, ICurrencyBucket> _currencyTradesPairs = new ConcurrentDictionary<string, ICurrencyBucket>();
+    public SuspiciousDealDetector(IDataLoader dataLoader, IOptions<SuspiciousDealDetectorOptions> options, ILogger logger)
+    {
+        _dataLoader = dataLoader;
+        _suspiciousDealDetectorOptions = options.Value;
+        _logger = logger;
+    }
+
+    public async Task<List<Trade>> LoadDealsAsync()
+    {
+        return await _dataLoader.LoadAllTradesAsync();
+    }
+
+    public async Task<List<List<Trade>>> DetectSuspiciousDealsAsync(List<Trade> trades)
+    {
+        List<List<Trade>> dealsAlreadyFilteredByCurrencyPairAndTimeStamp = new();
+
+        await SortTradesByCurrencyPairsAsync(trades);
+
+        var dealsInBuckets = _currencyTradesPairs.Values;
+
+        Log.Information($"Filtering trades into currency buckets. \n" +
+            $"{dealsInBuckets.Count} buckets found.");
+
+        foreach (var bucket in dealsInBuckets)
+        {
+            var tradesFromOneBucket = bucket.Trades;
+
+            var filteredGroups = GroupTradesByTimestampWithTimeTolerance(tradesFromOneBucket, SuspiciousDealDetector.TimeDifferTolerance);
+
+            dealsAlreadyFilteredByCurrencyPairAndTimeStamp.AddRange(filteredGroups);
+        }
+
+        Log.Information($"Filtering by Timestamp. " +
+            $"{dealsAlreadyFilteredByCurrencyPairAndTimeStamp.Count} buckets remains. {TradesTotalNumber(dealsAlreadyFilteredByCurrencyPairAndTimeStamp)} trades total. \n");
+
+        if(dealsAlreadyFilteredByCurrencyPairAndTimeStamp.Count > 0)
+        {
+            var dealsAlreadyFilteredByCurrencyPairAndTimeStampAndAction = TradeActionFilter(dealsAlreadyFilteredByCurrencyPairAndTimeStamp);
+
+            Log.Information($"Filtering by Action. {TradesTotalNumber(dealsAlreadyFilteredByCurrencyPairAndTimeStampAndAction)} trades total");
+
+            foreach(var deals in dealsAlreadyFilteredByCurrencyPairAndTimeStampAndAction)
+            {
+                var finalFilter = differenceInVolumeToBalanceRatioFilter(deals);
+            }
+
+            Log.Information($"Final filtering by Volume to balance ratio. {TradesTotalNumber(dealsAlreadyFilteredByCurrencyPairAndTimeStampAndAction)} trades total");
+
+            dealsAlreadyFilteredByCurrencyPairAndTimeStampAndAction.RemoveAll(list => !list.Any());
+
+            SuspiciousDealsLogVisulizer(dealsAlreadyFilteredByCurrencyPairAndTimeStampAndAction);
+
+            return dealsAlreadyFilteredByCurrencyPairAndTimeStampAndAction;
+
+        }
+
+        return dealsAlreadyFilteredByCurrencyPairAndTimeStamp;
+    }
+
+    #region Currency Bucket Sort
+    public async Task SortTradesByCurrencyPairsAsync(List<Trade> trades)
+    {
+
+        foreach (var trade in trades)
+        {
+            if (!_currencyTradesPairs.ContainsKey(trade.Currency))
+            {
+                _currencyTradesPairs.TryAdd(trade.Currency, new CurrencyBucket(trade.Currency));
```

Comment:
This is not thread safe way how to add or update because it is no atomic; two threads could jump into this condition but only one would create the new bucket while the second one would fail.

Try this instead

```
                _currencyTradesPairs.AddOrUpdate(trade.Currency, new CurrencyBucket(trade.Currency), (currency, bucket) =>
                {
                    bucket.Trades.Add(trade);

                    return bucket;
                }); 
```

Reply:
Read about the method and about what it does

Reply:
Done.